### PR TITLE
configuration added for HTTP Client timeouts

### LINKED ISSUE #1316: Support configuration of HTTP Client timeouts
**Is your feature request related to a problem? Please describe.**
The reliant client app times out while trying to sync data from some fhir server endpoints e.g. _https://fhir.xyz.xyz.org/fhir/MedicationRequest?subject.organization=361&_count=100&_sort=_lastUpdated_

**Describe the solution you'd like**
The HTTP client currently has a default 10 seconds configured for the various timeouts. We need to support configuration of the timeouts for the http client e.g. `connect timeout`, `read timeout`

**Describe alternatives you've considered**

Two approaches to solve this could be:

1. Allow configuration of the above individual timeout settings 
2. Allow configuration of a custom http client instance

**Additional context**
See screenshot for list of endpoints that throw a `socket timeout exception`

Screenshot 2022-04-21 at 10 21 05

**Would you like to work on the issue?**

### PR DESCRIPTION
**Pull request by PallaviGanorkar**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1316

**Description**
connectTimeout, readTimeout, writeTimeout configuration added for HTTP client 

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Feature

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Review by vitorpamplona — APPROVED** [review]

File: engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt [hunk 1]
```diff
@@ -127,3 +129,13 @@ enum class DatabaseErrorStrategy {
  * auth token that may be necessary to communicate with the server.
  */
 data class ServerConfiguration(val baseUrl: String, val authenticator: Authenticator? = null)
+
+/**
+ * A configuration to provide the network connectTimeout, readTimeout and writeTimeout in seconds
+ * while communicating with the server.
+ */
+data class NetworkConfiguration(
+  val connectionTimeOut: Long = 10,
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt:138]:
```suggestion
/** Connection timeout (in seconds). The default is 10 seconds. */
  val connectionTimeOut: Long = 10,
```

File: engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt [hunk 2]
```diff
@@ -127,3 +129,13 @@ enum class DatabaseErrorStrategy {
...
+  val readTimeOut: Long = 10,
+  val writeTimeOut: Long = 10
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt:140]: Add individual kdocs.

File: engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt [hunk 3]
```diff
@@ -127,3 +129,13 @@ enum class DatabaseErrorStrategy {
  * auth token that may be necessary to communicate with the server.
  */
 data class ServerConfiguration(val baseUrl: String, val authenticator: Authenticator? = null)
+
+/**
+ * A configuration to provide the network connectTimeout, readTimeout and writeTimeout in seconds
+ * while communicating with the server.
+ */
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt:136]:
```suggestion
/** A configuration to provide the network connection parameters.*/
```

File: engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt [hunk 4]
```diff
@@ -127,3 +129,13 @@ enum class DatabaseErrorStrategy {
...
+data class NetworkConfiguration(
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt:137]: Would it make sense for this to be part of `ServerConfiguration` ? @jingtang10 
Like 
```
ServerConfiguration(
  val baseUrl: String, 
  val networkConfiguration NetworkConfiguration = NetworkConfiguration(),
  val authenticator: Authenticator? = null
) 
```

File: engine/src/main/java/com/google/android/fhir/FhirServices.kt [hunk 5]
```diff
@@ -40,6 +40,7 @@ internal data class FhirServices(
     private var enableEncryption: Boolean = false
     private var databaseErrorStrategy = DatabaseErrorStrategy.UNSPECIFIED
     private var serverConfiguration: ServerConfiguration? = null
+    private var networkConfiguration: NetworkConfiguration = NetworkConfiguration()
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/FhirServices.kt:43]: Let's keep it nullable and pass a default value during `build()` if its not set.

File: engine/src/main/java/com/google/android/fhir/sync/remote/RemoteFhirService.kt [hunk 6]
```diff
@@ -38,7 +39,12 @@ internal interface RemoteFhirService : DataSource {
 
   @POST(".") override suspend fun upload(@Body bundle: Bundle): Resource
 
-  class Builder(private val baseUrl: String) {
+  class Builder(
+    private val baseUrl: String,
+    private val connectionTimeOut: Long,
+    private val readTimeOut: Long,
+    private val writeTimeOut: Long
+  ) {
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/sync/remote/RemoteFhirService.kt:47]:
```suggestion
  class Builder(
    private val baseUrl: String
  ) {
  
  fun setNetworkConfiguration(networkConfiguration NetworkConfiguration) {}
```

**Review by aditya-07 — CHANGES_REQUESTED** [review]

File: engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt (hunk 2 above)
Comment (PallaviGanorkar) [reply engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt:140]: Done

File: engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt (hunk 3 above)
Comment (PallaviGanorkar) [reply engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt:136]: Done

File: engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt (hunk 4 above)
Comment (jingtang10) [reply engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt:137]: i think this makes sense

Comment (PallaviGanorkar) [reply engine/src/main/java/com/google/android/fhir/FhirEngineProvider.kt:134]: DOne

**Review by aditya-07 — APPROVED** [review]
LGTM!

**Pull request merged**